bookshelf/src/CloudsqlModel.php

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * Class CloudsqlModel implements the DataModel with a mysql database.
+ *
+ * Set the three environment variables MYSQL_DSN, MYSQL_USER,
+ * and MYSQL_PASSWORD.
+ */
+class CloudsqlModel implements DataModelInterface
+{
+    /**
+     * Creates a new PDO instance and sets error mode to exception.
+     *
+     * @return PDO
+     */
+    private function newConnection()
+    {
+        $pdo = new PDO(getenv('MYSQL_DSN'), getenv('MYSQL_USER'),
+            getenv('MYSQL_PASSWORD'));
+        $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
+
+        return $pdo;
+    }
+
+    /**
+     * Creates the SQL books table if it doesn't already exist.
+     */
+    public function __construct()
+    {
+        $columns = array(
+            'id INT NOT NULL AUTO_INCREMENT PRIMARY KEY ',
+            'title VARCHAR(255)',
+            'author VARCHAR(255)',
+            'publishedDate DATE',
+            'imageUrl VARCHAR(255)',
+            'description VARCHAR(255)',
+            'createdBy VARCHAR(255)',
+            'createdById VARCHAR(255)',
+        );
+        $this->columnNames = array_map(function ($columnDefinition) {
+            return explode(' ', $columnDefinition)[0];
+        }, $columns);
+        $columnText = implode(', ', $columns);
+        $pdo = $this->newConnection();
+        $pdo->query("CREATE TABLE IF NOT EXISTS books ($columnText)");
+    }
+
+    public function listBooks($limit = 10, $cursor = null)
+    {
+        $pdo = $this->newConnection();
+        if ($cursor) {
+            $query = 'SELECT * FROM books WHERE id > :cursor ORDER BY id'.
+                ' LIMIT :limit';
+            $statement = $pdo->prepare($query);
+            $statement->bindValue(':cursor', $cursor, PDO::PARAM_INT);
+        } else {
+            $query = 'SELECT * FROM books ORDER BY id LIMIT :limit';
+            $statement = $pdo->prepare($query);
+        }
+        $statement->bindValue(':limit', $limit + 1, PDO::PARAM_INT);
+        $statement->execute();
+        $rows = array();
+        $last_row = null;
+        $new_cursor = null;
+        while ($row = $statement->fetch(PDO::FETCH_ASSOC)) {
+            if (count($rows) == $limit) {
+                $new_cursor = $last_row['id'];
+                break;
+            }
+            array_push($rows, $row);
+            $last_row = $row;
+        }
+
+        return array(
+            'books' => $rows,
+            'cursor' => $new_cursor,
+        );
+    }
+
+    public function create($book, $id = null)
+    {
+        $names = '';
+        $values = [];
+        $question_marks = '';
+        $separator = '';
+        $pdo = $this->newConnection();
+        foreach ($this->columnNames as $column) {
+            if (array_key_exists($column, $book)) {
+                $question_marks = "$question_marks$separator?";
+                $names = "$names$separator$column";
+                $separator = ', ';
+                array_push($values, $book[$column]);
```

Comment:
For these, I would prefer to see

```php
foreach(...) {
    if (...) {
        $question_marks[] = '?';
        $names[] = $column
        $values[] = $book[$column];
    }
}
$sql = sprintf('INSERT INTO books (%s) VALUES(%s)',
    implode(', ', $names),
    implode(', ', $query_marks)
);
//..
```

Or even better, avoid the `foreach` all together and do something like this:

```php
$names = array_keys($books);
$question_marks = array_fill(0, count($books), '?');
$sql = sprintf('INSERT INTO books (%s) VALUES(%s)',
    implode(', ', $names),
    implode(', ', $query_marks)
);
$statement = $pdo->prepare($sql);
$statement->execute($book);
```